src/CityOnPlanet.cpp

```diff
@@ -202,123 +312,345 @@ CityOnPlanet::~CityOnPlanet()
+	m_cityType = &s_cityFlavours[m_rand.Int32(s_cityFlavours.size())];
 
-	Random rand;
-	rand.seed(seed);
+	CalcCityRadius(m_body);
 
-	int population = planet->GetSystemBody()->GetPopulation();
-	int cityradius;
+	// TODO: run in a task thread
+	Generate(station);
 
-	if (planet->GetSystemBody()->HasAtmosphere()) {
-		population *= 1000;
-		cityradius = (population < 200) ? 200 : ((population > START_SEG_SIZE) ? START_SEG_SIZE : population);
-	} else {
-		population *= 100;
-		cityradius = (population < 250) ? 250 : ((population > START_SEG_SIZE_NO_ATMO) ? START_SEG_SIZE_NO_ATMO : population);
-	}
-
-	citybuildinglist_t *buildings = &s_buildingList;
-	vector3d cent = p;
-	const int cellsize_i = 80;
-	const double cellsize = double(cellsize_i);						// current widest building = 92
-	const double bodyradius = planet->GetSystemBody()->GetRadius(); // cache for bodyradius value
-
-	static const int gmid = (cityradius / cellsize_i);
-	static const int gsize = gmid * 2;
-
-	assert((START_SEG_SIZE / cellsize_i) < 100);
-	assert((START_SEG_SIZE_NO_ATMO / cellsize_i) < 100);
-	uint8_t cellgrid[200][200];
-	std::memset(cellgrid, 0, sizeof(cellgrid));
-
-	// calculate the size of the station model
-	const int x1 = floor(aabb.min.x / cellsize);
-	const int x2 = ceil(aabb.max.x / cellsize);
-	const int z1 = floor(aabb.min.z / cellsize);
-	const int z2 = ceil(aabb.max.z / cellsize);
-
-	// Clear the cells where the station is
-	for (int x = 0; x <= gsize; x++) {
-		for (int z = 0; z <= gsize; z++) {
-			const int zz = z - gmid;
-			const int xx = x - gmid;
-			if (zz > z1 && zz < z2 && xx > x1 && xx < x2)
-				cellgrid[x][z] = 1;
-		}
+	AddStaticGeomsToCollisionSpace();
+}
+
+void CityOnPlanet::Generate(SpaceStation *station)
+{
+	PROFILE_SCOPED()
+	Profiler::Clock _genTimer;
+	_genTimer.Start();
+
+	// Create the acceleration grid structure
+	// ==========================================
+
+	// Retrieve the station parameters
+	uint8_t stationSize[2] = {};
+	GetModelSize(station->GetAabb(), stationSize);
+
+	uint32_t stationSizeMax = std::max<uint32_t>(stationSize[0] / 2, stationSize[1] / 2);
+
+	// update the allowed radius to ensure the spaceport isn't larger than the entire city
+	m_cityRadius += stationSizeMax * CELLSIZE;
+
+	// For now, just assume the city is a circle around the spaceport in the center
+	const uint32_t cityExtents = (m_cityRadius / CELLSIZE);
+	m_citySize = cityExtents * 2;
+	m_gridPitch = std::ceil(m_citySize / 8.0);
+
+	Log::Verbose("Generating City for spacestation {}", station->GetSystemBody()->GetName());
+	Log::Verbose("\tpopulation: {0} size {1}x{1}c (radius {2})",
+		m_body->GetPopulation(), m_citySize, m_cityRadius);
+
+
+	// ensure divisible by 8 bytes for fast 64-bit lookup
+	if (m_gridPitch & 7) {
+		m_gridPitch = (m_gridPitch + 7) & ~7;
 	}
+	m_gridLen = m_gridPitch * m_citySize;
+
+	m_gridBitset.reset(new uint8_t[m_gridLen]);
+	std::memset(m_gridBitset.get(), 0, m_gridLen);
```

Comment:
now I don't understand why we don't use `std::vector`

Reply:
std::vector is a good fit for resizable arrays or piece-wise construction of contiguous lists. This is a single allocate-once slab of memory; `std::vector<bool>` is a nightmare to work with, `std::bitset` requires you to know the size of your allocation up-front, and given that we're reinterpreting the memory to enable aligned addressing anyways, just a raw memory pointer is good enough for this.

All of the access to m_gridBitset is hidden behind helper functions anyways, so a std::vector would not improve the code in any significant fashion other than RAII deletion of the memory which std::unique_ptr already does.